### ISSUE TITLE
Aleph's key trade dialog disappears

### ISSUE THREAD

**Issue by AmProsius** [AmProsius/gothic-1-community-patch#189]
### See [fix suggestion](#issuecomment-827498255). It will be fixed in #40.

----

**Describe the bug**
Aleph's dialog choice about the key trade disappears after loading a savegame, if the player didn't have enough ore.

**Expected behavior**
Aleph's dialog choice about the key trade no longer disappears, if the player didn't have enough ore.

**Steps to reproduce the issue**
1. Offer Aleph more than 15 ore for the key to the chests in the storage shed without enough ore in the inventory.
2. Save the game.
3. Load the save.
4. The dialog choice disappeared.

**Additional context**
https://github.com/AmProsius/gothic-1-community-patch/blob/839bd35f840802afb4884d69a79a7df3bfa050f9/scriptbase/_work/Data/Scripts/Content/Story/Missions/DIA_VLK_585_Aleph.d#L196-L290

**Comment by AmProsius** [comment]
Part of the solution might be to change
https://github.com/AmProsius/gothic-1-community-patch/blob/839bd35f840802afb4884d69a79a7df3bfa050f9/scriptbase/_work/Data/Scripts/Content/Story/Missions/DIA_VLK_585_Aleph.d#L196-L204

to

```d
instance VLK_585_Aleph_SCHUPPEN (C_INFO)
{
    npc             = VLK_585_Aleph;
    condition       = VLK_585_Aleph_SCHUPPEN_Condition;
    information     = VLK_585_Aleph_SCHUPPEN_Info;
    important       = 0;
    permanent       = 1;
    description     = "Do you know anything about the key to the chests in the storage shed??";
};
```

and

https://github.com/AmProsius/gothic-1-community-patch/blob/839bd35f840802afb4884d69a79a7df3bfa050f9/scriptbase/_work/Data/Scripts/Content/Story/Missions/DIA_VLK_585_Aleph.d#L205-L211

to

```d
FUNC int  VLK_585_Aleph_SCHUPPEN_Condition()
{
    if ( Npc_KnowsInfo (hero, GRD_271_ULBERT_DRUNK) ) 
    && ( !Npc_KnowsInfo (hero, GRD_261_Brandick_ALEPH) )
    && ( !Npc_HasItems(hero, ItKe_OM_03) )
    {
        return 1;
    };
};
```

but we will also have to take into account that if the player gave Aleph only 15 ore, the dialog choice should also disappear.

https://github.com/AmProsius/gothic-1-community-patch/blob/839bd35f840802afb4884d69a79a7df3bfa050f9/scriptbase/_work/Data/Scripts/Content/Story/Missions/DIA_VLK_585_Aleph.d#L280-L285

Unfortunately I don't know yet how to check if that particular dialog was already spoken.

**Comment by szapp** [comment]
These dialogs confuses me. As you state in the bottom of the comment above, it looks like the dialog will no longer be available if the player offers only 15 ore. However, if I am not mistaken, this dialog choice does not clear the remaining choices, such that the player will return to choose from another dialog choice. The only way to leave the dialog choices then, is by either buying the key or selecting "back", in which case the dialog is set to permanent again. To me it seems that the dialog will thus always be available until the player obtains the key (of course only until saving and loading). Even weirder is the dialog `VLK_585_Aleph_DIRTY` in which the player has two choices, to pay or not to pay. If the player does not have 100 ore and chooses not to pay, the dialog is set to non-permanent, but the player still has the choice to pay in which case there is not enough ore and the dialog will be set to permanent and remain. If the player does this in the opposite order, the dialog is gone. I might be wrong about this, since I have not confirmed this in game. It seems Aleph's dialogs are completely broken.

To me it seems there are two mistakes in the dialog of this issue then:
1. The dialog will disappear on saving and loading (as described by this issue).
2. The dialog was intended to vanish if the player blew their chance (15 ore), but the dialog will in fact remain (until saving and loading).

I think we can neglect the second mistake and assume that the player always has the chance to get the key (even when first offering only 15 ore). If we can agree on that, the fix will be possible:

In all cases, when the player obtains the key, a log entry will be written:
https://github.com/AmProsius/gothic-1-community-patch/blob/839bd35f840802afb4884d69a79a7df3bfa050f9/scriptbase/_work/Data/Scripts/Content/Story/Missions/DIA_VLK_585_Aleph.d#L190-L194

We can just make the respective dialogs permanent (implies we will need a revertible fix) and extend the dialog conditions to check if the log entry exists, which is written in `B_Aleph_StorageShedKey` (implies we will need to infer the string of the log entry from the bytecode).

This will be a bit of work.

### Once per session (first time applying the fix):
1. Check if the infos `VLK_585_Aleph_SCHUPPEN` and `VLK_585_Aleph_DIRTY` exist
2. Check that the infos `VLK_585_Aleph_SCHUPPEN` and `VLK_585_Aleph_DIRTY` are non-permanent (to ensure the fix is not already handled by someone else)
3. Check if function `B_Aleph_StorageShedKey` exists
4. Check if string constant `CH2_StorageShed` exists
5. Search the bytecode of `B_Aleph_StorageShedKey` to find and remember the log entry string
6. Check if the functions `VLK_585_Aleph_SCHUPPEN_30`, `VLK_585_Aleph_SCHUPPEN_50` and `VLK_585_Aleph_DIRTY_100` exist
7. Check if the function `B_Aleph_StorageShedKey` is called from `VLK_585_Aleph_SCHUPPEN_50`, `VLK_585_Aleph_SCHUPPEN_30` and `VLK_585_Aleph_DIRTY_100`
8. If all above is given, add this to the dialog conditions.
https://github.com/AmProsius/gothic-1-community-patch/blob/839bd35f840802afb4884d69a79a7df3bfa050f9/scriptbase/_work/Data/Scripts/Content/Story/Missions/DIA_VLK_585_Aleph.d#L205-L211
changed to
```D
FUNC int  VLK_585_Aleph_SCHUPPEN_Condition()
{
    if (( Npc_KnowsInfo (hero, GRD_271_ULBERT_DRUNK) ) && ( !Npc_KnowsInfo (hero, GRD_261_Brandick_ALEPH) ))
    && (G1CP_LogTopicHasEntry(CH2_StorageShed, "...")) // As infered from before
    {
        return 1;
    };
};
```
and
https://github.com/AmProsius/gothic-1-community-patch/blob/839bd35f840802afb4884d69a79a7df3bfa050f9/scriptbase/_work/Data/Scripts/Content/Story/Missions/DIA_VLK_585_Aleph.d#L303-L309
changed to
```D
FUNC int  VLK_585_Aleph_DIRTY_Condition()
{
    if (( Npc_KnowsInfo (hero, GRD_271_ULBERT_DRUNK) ) && ( Npc_KnowsInfo (hero, GRD_261_Brandick_ALEPH) ))
    && (G1CP_LogTopicHasEntry(CH2_StorageShed, "...")) // As infered from before
    {
        return 1;
    };
};
```

### Apply and revert every time
Make the infos `VLK_585_Aleph_SCHUPPEN` and `VLK_585_Aleph_DIRTY` permanent.

**Comment by szapp** [comment]
If the fix is not implemented as suggested above, #190 should be reopened.

**Comment by szapp** [comment]
If the fix is addressed as suggested above, it may be wise to close this issue and fix it through #40 as suggested above.

**Comment by AmProsius** [comment]
This issue will be referenced in #40 as suggested.
